src/endpoint.jl

```diff
@@ -25,6 +26,27 @@ struct Endpoint{T, D, B <: Bound}
 end
 
 Endpoint{T,D,B}(ep) where {T, D, B <: Bounded} = Endpoint{T,D,B}(convert(T, ep))
+Base.eltype(::AbstractEndpoint{T}) where T = T
+Base.eltype(::Type{<:AbstractEndpoint{T}}) where T = T
+
+# when all intervals are half-closed in the same direction (all [a, b) or all
+# (b, a]) set operations over them are closed (e.g. all resulting intervals will
+# be the same type). In these cases we can improve type invariance by using a
+# `HalfOpenEndpoint`, which indicates whether the bounds are left-closed or
+# right-closed, storing the left/right-ness as a flag
+struct DirectionBound{T} end
+const LeftClosed = DirectionBound{:LeftClosed}()
+const RightClosed = DirectionBound{:RightClosed}()
+struct HalfOpenEndpoint{T, B} <: AbstractEndpoint{T}
```

Comment:
Intervals are half-open/closed and not endpoints of intervals. This seems to me like maybe you want is a `HalfOpenInterval` type. I'll need to take a closer look into how this is used though

Reply:
The goal is as follows: during the process of "unbunching" intervals into endpoints, and then "bunching" them back into intervals, I want to preserve the fact that the original array has an all left-closed or all right-closed intervals. That's because 
1.) said form leads to type stable arrays  
2.) said form can ignore the checks for edge types in `mergesets` (hence the ).

I guess another approach could be to have a flag and/or special container that denotes the
"all closed in one direction" property, and use the existing Endpoint type in all cases.

Reply:
The problem with the design as I see it is that endpoints are dealing with the left/right side of the interval without knowing anything about the other endpoint in the interval. I'll be able to speak better to alternatives once I read through the rest of the code.

Reply:
> It's not obvious to me, at the moment, how to write that in a type stable way though.

I'd suggest putting code readability over type stability to start with as otherwise you can optimize yourself into a corner

Reply:
> The problem with the design as I see it is that endpoints are dealing with the left/right side of the interval without knowing anything about the other endpoint in the interval

My idea was to create a subtype of endpoint that is only used in those cases where you do know something about the other endpoints. I do also feel that there is probably a better design, combing back to this now (after some time away). I'll think it over some more and wait for whatever further comments you have.